src/base59.cpp

```diff
@@ -0,0 +1,150 @@
+#include <assert.h>
+#include <limits>
+#include <stdexcept>
+#include <string.h>
+#include <vector>
+#include <veriblock/base58.hpp>
+#include <string>
+
+namespace VeriBlock {
+
+static std::string g_Base59Alphabet =
+    "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz0";
+static std::vector<uint32_t> g_Indexes(128);
+static const uint32_t g_kBase_256 = 256;
+static const size_t g_kBase59 = g_Base59Alphabet.size();
+
+static uint8_t divmod59(std::vector<uint8_t> &number, size_t startAt) {
+    size_t remainder = 0;
+  for (size_t i = startAt; i < number.size(); i++) {
+    uint32_t digit256 = (uint32_t)number[i] & 0xFF;
+    size_t temp = remainder * g_kBase_256 + digit256;
+    number[i] = (uint8_t)(temp / g_kBase59);
+    remainder = temp % g_kBase59;
+  }
+  return (uint8_t)remainder;
+}
+
+static uint8_t divmod256(std::vector<uint8_t> &number59, size_t startAt) {
+    size_t remainder = 0;
+  for (size_t i = startAt; i < number59.size(); i++) {
+    int digit59 = (int8_t)number59[i] & 0xFF;
+    size_t temp = remainder * g_kBase59 + digit59;
+    number59[i] = (uint8_t)(temp / g_kBase_256);
+    remainder = temp % g_kBase_256;
+  }
+  return (uint8_t)remainder;
+}
+
+std::string EncodeBase59(const void *buf, size_t nSize) {
+  std::string result;
+
+  if (!nSize) {
+    return result;
+  }
+
+  // Make a copy of the input since we are going to modify it.
+  std::vector<uint8_t> input(nSize);
+  if (input.empty()) {
+    return result;
+  }
+  memcpy(input.data(), buf, nSize);
+
+  // Count leading zeroes
+  size_t zeroCount = 0;
+  while (zeroCount < input.size() && input[zeroCount] == 0) {
+    ++zeroCount;
+  }
+
+  // The actual encoding
+  std::vector<uint8_t> temp(input.size() * 2);
+  size_t j = temp.size();
+
+  size_t startAt = zeroCount;
+  while (startAt < input.size()) {
+    uint8_t mod = divmod59(input, startAt);
+    if (input[startAt] == 0) {
+      ++startAt;
+    }
+    temp[--j] = g_Base59Alphabet[mod];
+  };
+
+  // Strip extra '1' if any
+  while (j < temp.size() && temp[j] == g_Base59Alphabet.c_str()[0]) {
+    ++j;
+  }
+
+  // Add as many leading '1' as there were leading zeros.
+  for (--zeroCount; zeroCount != (size_t)-1; zeroCount--) {
+      temp[--j] = g_Base59Alphabet[0];
+  }
+
+  result.resize(temp.size() - j);
+  memcpy((void *)result.data(), &temp.data()[j], temp.size() - j);
+  return result;
+};
+
+std::string EncodeBase59(const unsigned char *pbegin, 
+    const unsigned char *pend) {
+    return EncodeBase59(pbegin, pend - pbegin);
+};
+
+std::vector<uint8_t> DecodeBase59(const std::string &input) {
+  std::vector<uint8_t> result;
+  if (input.empty()) {
+    return result;
+  }
+
+  memset(g_Indexes.data(), -1, g_Indexes.size() * sizeof(uint32_t));
```

Comment:
try to not use `memcpy`, `memset` and friends. use vector's constructors

also, why `g_Indexes` is not function-local?
when you declare it outside, you make this function thread-unsafe, because you do **writes** to global array

Reply:
Porting artefact. Agree.

Reply:
fixed